ensemble/randomforest.h

```diff
@@ -0,0 +1,53 @@
+#ifndef _MACHINELEARNING_ENSEMBLE_RANDOMFOREST_H_
+#define _MACHINELEARNING_ENSEMBLE_RANDOMFOREST_H_
+
+#include "../tree/tree.h"
+
+namespace ensemble {
+/**
+ * Base class for Random Forest Classification and Regression algorithms.
+ */
+class RandomForest {
+ public:
+  /**
+   * \brief Standard destructor.
+   */
+  virtual ~RandomForest();
+
+  /**
+   * Compute and return the score of the training dataset obtained using an
+   * out-of-bag estimate
+   * \return oob score
+   */
+  virtual double OobScore() = 0;
+
+  /*
+   * Build the RandomForest
+   * \param thread Number of threads for paralel execution
+   * \param tree Number of trees in the Forest
+   */
+  virtual void Build(/* Collection col, */ const int thread,
+                     const int tree) = 0;
+
+  /**
+   * Predicts the labels of the instances of X based on a vote by
+   * the trees of the forest, weighted by their probability estimates
+   * \param Instance labels
+   */
+  virtual void Predict(/* Collection col, */
```

Comment:
I'd actually implement the prediction of a single sample before we worry about multi-sample predictions. It feels more natural and will eventually be used on multi-sample prediction.